### ISSUE TITLE
Hard-coded fips packages will not work on bionic

### ISSUE THREAD

**Issue by cmars** [canonical/ubuntu-advantage-client#1050]
# BLOCKED on a FIPS metapackage from security/kernel team which expresses conditional pkg dependencies which can't be expressed in debian/control

Archive repositories for fips and fips-updates are now available on esm.ubuntu.com but the client will not be able to enable them (apt keyring and grub issues aside). The next problem you'll encounter there is the client is hard-coded[0] to install libssl-1.0.0 fips packages, but the package is libssl-1.1 on bionic.

Contracts can be updated to provide the availablePackages directive per series, with the right package names for the series you're on, but the client will need to be updated to use it.

[0] https://github.com/canonical/ubuntu-advantage-client/blob/7f61778a72f749b75016ae34163ba0239a4a5527/uaclient/entitlements/fips.py#L16

**Comment by blackboxsw** [comment]
@cmars thanks for this issue. FIPS support ultimately won't be published until we get a single fips* meta package that manages all fips package version dependencies directly in its own debian/control file. This work will be related to #1027.   I'd like UA Client to rely directly on dynamic package versions exposed by UA contract server's additionalPackages directive for this work.

**Comment by blackboxsw** [comment]
Blocked also on ua-contract work to define a more full schema definition for additionalPackages 
https://github.com/CanonicalLtd/ua-contracts/issues/851

**Comment by kikoreis** [comment]
Can we document a workaround that can unblock the client from at least granting access to the repository?

**Comment by mitechie** [comment]
Yes, @blackboxsw will provide manual instructions for attaching that bypasses the client while we get the deps updated in the codebase to support the different series.

**Comment by blackboxsw** [comment]
@kikoreis @cmars @mitechie 
Here is a script that can be run on and Azure Ubuntu PRO host to enable fips against staging.
@cmars is in process of making that contract available on production, so we could drop the 3 lines
```shell
ua detach
sed -i...
ua auto-attach
```

```bash
#!/bin/bash
# manual fips enablement

[ "$(whoami)" != "root" ] && echo "Must be root" && exit 1

ua detach
sed -i 's/contracts.canonical.com/contracts.staging.canonical.com/' /etc/ubuntu-advantage/uaclient.conf
ua auto-attach

machine_token=$(cat /var/lib/ubuntu-advantage/private/machine-token.json | json_pp | awk -F '"' '/machineToken" \:/{print $4}')

cat >> /etc/apt/auth.conf.d/90ubuntu-advantage <<EOF
machine esm.staging.ubuntu.com/fips/ login bearer password ${machine_token} # manual fips preview credentials" >> /etc/apt/auth.conf.d/90ubuntu-advantage
EOF

cat > /etc/apt/sources.list.d/ubuntu-fips-bionic.list <<EOF
deb https://esm.staging.ubuntu.com/fips/ubuntu bionic main
# deb-src https://esm.staging.ubuntu.com/fips/ubuntu bionic main
EOF

cat > /etc/apt/preferences.d/ubuntu-fips-bionic <<EOF
Package: *
Pin: release o=UbuntuFIPS, n=bionic
Pin-Priority: 1001
EOF

apt-key adv --keyserver keyserver.ubuntu.com --recv-key e23341b2a1467edbf07057d6c1997c40ede22758

apt-get update

PKGS="fips-initramfs linux-fips"

for pkg in openssl libssl1.1 openssh-client openssh-server strongswan; do
   found=$(dpkg-query --show $pkg 2> /dev/null)
   if [ -n "$found" ]; then
       # install fips versions and -hmac
       PKGS="${PKGS} $pkg"
       found_hmac=$(apt-cache policy $pkg-hmac 2> /dev/null)
       if [ -n "$found_hmac" ]; then
           PKGS="${PKGS} $pkg-hmac"
       fi
   fi
done

apt-get install $PKGS


fips_grub_boot_option=$(awk -F "'" '/menuentry.*fips-advanced/{print $(NF-1)}' /boot/grub/grub.cfg)

cat > /etc/default/grub.d/99-fips.cfg <<EOF
GRUB_DEFAULT="1>${fips_grub_boot_option}"
GRUB_CMDLINE_LINUX_DEFAULT="\$GRUB_CMDLINE_LINUX_DEFAULT fips=1"
EOF
update-grub

echo "Please reboot the server to enable fips certified kernel"
```


Once @cmars updates production contracts

**Comment by cmars** [comment]
@kikoreis @blackboxsw Production contracts are updated. If you remove the detach/sed/auto-attach and replace `esm.staging.canonical.com` with `esm.canonical.com` in the script above, works with Ubuntu 18.04 Pro in Azure.

**Comment by blackboxsw** [comment]
> @kikoreis @blackboxsw Production contracts are updated. If you remove the detach/sed/auto-attach and replace `esm.staging.canonical.com` with `esm.canonical.com` in the script above, works with Ubuntu 18.04 Pro in Azure.

Thanks @cmars I've updated  the script now.

# note: downreving the azure kernel 5.0  to fips 4.15 results in cloud-init seeing a new instance-id (because 4.15 capitalized the uuid and 5.0 uses lowercase). So cloud-init regenerates host keys as it sees a delta

My logs below:
```
# 5.0 kernel present in Azure PRO gets us this instance-id
16:55 2020-05-12 22:11:00,430 - main.py[DEBUG]: [local] init will now be targeting instance id: e9398239-f6ff-1145-bfac-f75568b82de4. new=True
# after downgrade to 4.15 fips kernel
16:55 2020-05-12 22:35:51,102 - main.py[DEBUG]: [local] init will now be targeting instance id: E9398239-F6FF-1145-BFAC-F75568B82DE4. new=True
```

**Comment by kikoreis** [comment]
Thanks for the workaround instructions. I had assumed we'd be able to use "ua enable fips" and just have that stop short of installing the FIPS packages as necessary, but I just realized that would make it inconsistent with how the command would work for 16.04 LTS, where it does in fact install all the relevant FIPS certified packages.

**Comment by blackboxsw** [comment]
ua-contracts continues to surface additionalPackages directive for repo-based services, even if the backend will not describe our conditional dependencies for the *-hmac deb packages.

@lucasmoura this work is probably two parts for UA Client:

1. we should be able to look at minimally refactoring RepoEntitlement.packages method to leverage existing additionalPackages directives:

Those directives are seen in /var/lib/ubuntu-advantage/private/machine-token.json on an attached machine

It might look something like following:
```
def packages(self) ....:
        self.cfg.entitlements.get(self.name, {})              
        directives = entitlement_cfg.get("entitlement", {}).get(                
            "directives", {}                                                    
        )                                                                       
        self.packages = directives.get("additionalPackages")
```

part 2 (as a separate pull request after we've talked about remaining FIPS work):
FIPS has an unique need in this space, it has only two required packages, and a number of conditional *-hmac packages which are only to be installed only if the corresponding base package is already installed in the running machine environment.

we'll need to wait on part 2 because @cmars mentioned in CanonicalLtd/ua-contracts#851 that the security team will be delivering a tool in the fips metapackage which will take care of those details for us.

**Comment by blackboxsw** [comment]
Fixed by #1064
